plugins/net_plugin/include/eosio/net_plugin/auto_bp_peering.hpp

```diff
@@ -0,0 +1,227 @@
+#pragma once
+
+#include <eosio/chain/producer_schedule.hpp>
+
+namespace eosio::auto_bp_peering {
+
+///
+/// This file implements the functionality for block producers automatically establishing p2p connections to their
+/// neighbors on the producer schedule.
+///
+
+using chain::account_name;
+using chain::flat_map;
+using chain::flat_set;
+
+template <typename Derived, typename Connection>
+class connection_manager {
+#ifdef BOOST_TEST_MODULE
+ public:
+#endif
+
+   flat_map<account_name, std::string> bp_peer_addresses;  // thread safe only because modified at plugin startup currently
+   flat_map<std::string, account_name> bp_peer_accounts;   // thread safe only because modified at plugin startup currently
+   flat_set<account_name>              my_accounts;
+   flat_set<account_name>              pending_neighbors;
+   flat_set<account_name>              active_neighbors;
+   uint32_t                            pending_schedule_version = 0;
+   uint32_t                            active_schedule_version  = 0;
+
+   Derived*       self() { return static_cast<Derived*>(this); }
+   const Derived* self() const { return static_cast<const Derived*>(this); }
+
+   template <template<typename...> typename Container, typename... Rest >
+   static std::string to_string(const Container<account_name, Rest...>& peers) {
+      std::stringstream result;
+      const char*       sep = "";
+      for (auto p : peers) {
+         result << sep << p.to_string();
+         sep = ", ";
+      }
+      return result.str();
+   }
+
+   class neighbor_finder_type {
+
+      const flat_map<account_name, std::string>&    bp_peer_addresses;
+      const flat_set<account_name>&                 my_accounts;
+      const std::vector<chain::producer_authority>& schedule;
+      chain::flat_set<std::size_t>                  indices;
```

Comment:
name `indices` doesn't sound self descriptive. I suggest to rename it to something like `my_schedule_indices` or something like that.